src/peft/tuners/tuners_utils.py

```diff
@@ -523,3 +526,33 @@ def inspect_matched_modules(tuner: BaseTuner, adapter_name: str = "default") ->
         else:
             module_dict["unmatched"].append(key)
     return module_dict
+
+
+def _maybe_include_all_linear_layers(peft_config: PeftConfig, model: nn.Module) -> PeftConfig:
+    """
+    Helper function to update `target_modules` to all linear/Conv1D layers if provided as 'ALL'. Adapted from the QLoRA
+    repository: https://github.com/artidoro/qlora/blob/main/qlora.py
+    """
+    if peft_config.target_modules == "ALL":
+        is_loaded_in_8bit = getattr(model, "is_loaded_in_8bit", False)
+        is_loaded_in_4bit = getattr(model, "is_loaded_in_4bit", False)
+        # match with a list of linear layer classes. this is needed as sometimes you can
+        # have a mix Eg. T5 with 8bit has instances of torch.nn.Linear and bnb.nn.Linear8bitLt
+        linear_classes = [torch.nn.Linear, Conv1D]
+        if is_bnb_available():
+            import bitsandbytes as bnb
+
+            linear_classes = [bnb.nn.Linear8bitLt] + linear_classes if is_loaded_in_8bit else linear_classes
+            linear_classes = [bnb.nn.Linear4bit] + linear_classes if is_loaded_in_4bit else linear_classes
+
+        linear_module_names = set()
+        for name, module in model.named_modules():
+            # match with all linear classes.
+            for cls in linear_classes:
+                if isinstance(module, cls):
+                    names = name.split(".")
+                    linear_module_names.add(names[0] if len(names) == 1 else names[-1])
+        if "lm_head" in linear_module_names:  # needed for 16-bit
```

Comment:
What about using this method: https://github.com/huggingface/transformers/blob/main/src/transformers/integrations/bitsandbytes.py#L267 ?

Reply:
Great idea! I didn't know about GPTNeoX's output layer! 

Regarding the `get_keys_to_not_convert` method: I don't think the entire method is useful, as it is meant for detecting keys/ child modules that  should not be quantized. With the `_maybe_include_all_linear_layers` function, we just need to make sure only internal transformer blocks are considered without the final classification layer if any.  I've copied over only the part that gets the key for the final output layer (also, I'm assuming only one such key will match - the `get_keys_to_not_convert` [returns a list](https://github.com/huggingface/transformers/blob/3b7675b2b844b02d4821b827871a21ad16dd446c/src/transformers/integrations/bitsandbytes.py#L295C13-L295C13) - I'm not aware of any architecture where there could be a clash - Correct me if i'm wrong).